### PR TITLE
drivers: sensor: ntc-thermistor fixes

### PR DESCRIPTION
**Pull request by panikiel**
This patchset addresses multiple problems with the ntc-thermistor driver.

### CODE REVIEW

File: drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c
Comment (asemjonovs) [inline drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c:39]: This looks like my original implementation :-)  

The Zephyr community advised me to use the bsearch for space optimization.  The compare function pointer here should find the lower key where the value falls between.  I don't think this change is needed. See
https://github.com/zephyrproject-rtos/zephyr/pull/54947#pullrequestreview-1335999939 and https://pastebin.com/prpRVmjn

File: drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c
Comment (asemjonovs) [inline drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c:70]: This is looking more like [linux implementation](https://github.com/torvalds/linux/blob/master/drivers/hwmon/ntc_thermistor.c#L382).  Lets add the additional safety checks here as well.

**Review by asemjonovs — CHANGES_REQUESTED** [review]

File: drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c
Comment (semihalf-anikiel-pawel) [reply drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c:39]: Oh, I wasn't aware that the bsearch() approach was requested during review. Thanks for the link.

However, I don't understand how this is space optimization. The comparison function still takes up code.
I compared the generated code with and without this change (ARM build):
```
without this change:
  ntc_compensation_compare_ohm(): 62 bytes
  ntc_get_temp_mc(): 114 bytes
  compare_ohm_DT_DRV_COMPATx(): 16 bytes (per instance)
with this change:
  ntc_get_temp_mc(): 132 bytes
```
Am I missing something?

Comment (asemjonovs) [reply drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c:39]: @cfriedt - looks like with implementing our own binary search algorithm we're saving space over using `bsearch`.  Any concerns with removing this?

Comment (cfriedt) [reply drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c:39]: There are a lot of changes happening in this PR. It would be ideal to make an apples to apples comparison.

One aspect of using bsearch is that the search itself is less intermixed with other code that tweaks this or that. So it's maybe a bit more obvious what is happening. So maybe code is a bit more maintainable / less bug prone?

Also, if 5 different drivers used bsearch instead of implementing their own bsearch, then there is commonality savings.

It is not very difficult to reduce code size for one custom use case without bsearch of course.

E.g. for small arrays, the code size could be tiny and a linear search would be faster

Comment (semihalf-anikiel-pawel) [reply drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c:40]: > One aspect of using bsearch is that the search itself is less intermixed with other code that tweaks this or that. So it's maybe a bit more obvious what is happening and code is maybe a bit more maintainable / less bug prone?

IMHO, this comparison function is much more complicated and error prone than a 7-line binary search implementation. 

> Also, if 5 different drivers used bsearch instead of implementing their own bsearch, then there is commonality savings.

That would happen if they wanted the find-exact behavior (and in that case yes, I agree using bsearch() is a good idea). But, if they want the find-largest-not-greater-than behavior, they would all have to implement the (again, IMO) overly complicated comparison function that "tricks" bsearch() into the other behavior.

> It is not very difficult to reduce code size for one custom use case without bsearch of course.

My intent with this change isn't to optimize code size (the size comparison I made was to question the original explanation for this approach), but to make the code more understandable than the original approach. If you disagree, I can keep the old one.

> E.g. for small arrays, the code size could be tiny and a linear search would be faster

Actually, is there a reason for not doing that? Are the compensation tables ever going to be really big?

File: drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c
Comment (semihalf-anikiel-pawel) [reply drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c:71]: Done

**Comment by semihalf-anikiel-pawel** [comment]
The Run tests failure seems to be unrelated to my changes? I didn't touch PM (nor does this driver use PM).

File: drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c
Comment (cfriedt) [reply drivers/sensor/ntc_thermistor/ntc_thermistor_calc.c:40]: I'm not a sensor maintainer, so will defer to them.

**Review by asemjonovs — DISMISSED** [review]

**Comment by semihalf-anikiel-pawel** [comment]
> Would you be able to fix the script as well? I was just made aware that this didn't get updated with an earlier PR. https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/utils/ntc_thermistor_table.py

Sure. I changed it to just output the zephyr,compensation-table property, it that fine?

File: scripts/utils/ntc_thermistor_table.py
Comment (asemjonovs) [inline scripts/utils/ntc_thermistor_table.py:52]: Lets keep this portion.  The idea is to create a DT node that one can copy/paste into their device tree file.  We can drop the `status` and `compatible` like you've already done.

Comment (semihalf-anikiel-pawel) [reply scripts/utils/ntc_thermistor_table.py]: Done

**Review by asemjonovs — DISMISSED** [review]

File: drivers/sensor/ntc_thermistor/ntc_thermistor.c
Comment (semihalf-anikiel-pawel) [reply drivers/sensor/ntc_thermistor/ntc_thermistor.c:44]: If `CONFIG_PM_DEVICE_RUNTIME` isn't defined, this function becomes a stub that returns zero, which should be optimized out: https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/pm/device_runtime.h#L179. Same for the `_put` below.

**Review by asemjonovs — DISMISSED** [review]

**Comment by semihalf-anikiel-pawel** [comment]
@asemjonovs Could you please take another look? I made some changes to the PM logic and that dismissed your approval.

**Comment by fabiobaltieri** [comment]
@semihalf-anikiel-pawel this needs a rebase as well btw

**Review by asemjonovs — DISMISSED** [review]

File: drivers/sensor/ntc_thermistor/ntc_thermistor.c
Comment (semihalf-anikiel-pawel) [reply drivers/sensor/ntc_thermistor/ntc_thermistor.c:108]: Yes, that's definitely better. Done

**Review by fabiobaltieri — DISMISSED** [review]

**Review by asemjonovs — DISMISSED** [review]

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Except for the DT_INST concern, the rest of the PR looks fine

File: drivers/sensor/ntc_thermistor/ntc_thermistor.c
Comment (semihalf-anikiel-pawel) [reply drivers/sensor/ntc_thermistor/ntc_thermistor.c:141]: I was looking at these two drivers that also declare multiple compatibles:
- `drivers/eeprom/eeprom_at2x.c`
- `drivers/display/ssd16xx.c`

I tried to be consistent with how they do it. What are your thoughts about this?

File: drivers/sensor/ntc_thermistor/ntc_thermistor.c
Comment (semihalf-anikiel-pawel) [reply drivers/sensor/ntc_thermistor/ntc_thermistor.c:50]: Should I put it into a seperate commit?

Comment (MaureenHelm) [reply drivers/sensor/ntc_thermistor/ntc_thermistor.c:50]: Yes

File: drivers/sensor/ntc_thermistor/ntc_thermistor.c
Comment (MaureenHelm) [reply drivers/sensor/ntc_thermistor/ntc_thermistor.c:141]: There are examples of both styles in the tree and some disagreement about which way to do it. Until there's a project-wide decision on a preferred style, I'd like to stick with `DT_INST` for now.

File: drivers/sensor/ntc_thermistor/ntc_thermistor.c
Comment (semihalf-anikiel-pawel) [reply drivers/sensor/ntc_thermistor/ntc_thermistor.c:50]: Done

File: drivers/sensor/ntc_thermistor/ntc_thermistor.c
Comment (semihalf-anikiel-pawel) [reply drivers/sensor/ntc_thermistor/ntc_thermistor.c]: Okay, I brought back the `DT_INST` style declarations.

**Review by MaureenHelm — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by asemjonovs — APPROVED** [review]

**Pull request merged**
